### ISSUE TITLE
Windnet: Forecast correct values?

### ISSUE THREAD

**Issue by peterdudfield** [openclimatefix/india-forecast-app#35]
## Describe the bug

I'm not totalyl convience the forecast is working proerply

## Expected behavior
follow the generation values

## Additional context
- might need to check the batch values are going in correctly
- might need to check the live ecmwf values are simiarl to the backtest ones. Sometimes they could be out by a factor of 1000 due to units
- are the NWP variables in the correct order?
- I had to rename `t`->`t2m` and `tlc` to `tcc`, is this right? linked with https://github.com/openclimatefix/nwp-consumer/issues/131

## Things done / TODO
 - [x] Updated location in database
 - [x] Check variables names are ok
 - [x] TCC looks right, so ok to rename from clt
 - [x] Check nwp are in the right order, yes ocf_datapipes does this. the xr_data.sel(...) does this
 - [x] update lon/ lat position
 - [x] Check normalize variables are the same
 - [x] Check live and backtest ecmwf are in 0.1 degrees
 - [x] Check wind speed are similar to [open meteo](https://open-meteo.com/en/docs/ecmwf-api/#hourly=wind_speed_10m,wind_direction_10m&wind_speed_unit=ms&start_date=2024-03-05&end_date=2024-03-09&time_mode=time_interval)
 - [ ] compare backtest batch with live batch
 - [x] Checked that NWP targettimes do change, when running app at different times
 - [ ] ...

**Comment by jacobbieker** [comment]
The `t` to `t2m` might be fine, I'll check that. But the `tlc` to `tcc` seems like it might be wrong. As that is setting the total low cloud cover to the total cloud cover

**Comment by jacobbieker** [comment]
Actually, I believe its because in the training data, the longitude is `74.6399` while in the prod example batch it is `72.6399`, which is from me moving it around to align more with where the wind sources are

**Comment by jacobbieker** [comment]
The variable renaming might also still be an issue, but that is less of one I think

**Comment by devsjc** [comment]
See https://github.com/openclimatefix/nwp-consumer/commit/6fa7fde9a658abde20064fd14cf6823c771b0e75 for handling the renaming. Updating the version of the consumer in terraform will propogate this.

**Comment by jacobbieker** [comment]
Here is where the code takes the metadata from the database and returns the lat/lon tuples for the dataloader: https://github.com/openclimatefix/india-forecast-app/blob/8d42a3d95cbeb9b6ffbeeaf8dde6681a2291f9ff/india_forecast_app/models/pvnet/model.py#L174C9-L181

**Comment by peterdudfield** [comment]
> Actually, I believe its because in the training data, the longitude is `74.6399` while in the prod example batch it is `72.6399`, which is from me moving it around to align more with where the wind sources are

And `26.4499` latitude?

**Comment by peterdudfield** [comment]
Does `time-of-day' go in as a feature?

**Comment by peterdudfield** [comment]
Could the model be trained with a weighting on the first few hours, to help reduce the error in the earl forecast horizons?

**Comment by jacobbieker** [comment]
> > Actually, I believe its because in the training data, the longitude is `74.6399` while in the prod example batch it is `72.6399`, which is from me moving it around to align more with where the wind sources are
> 
> And `26.4499` latitude? 

Yes, that latitude is correct

**Comment by jacobbieker** [comment]
> Does `time-of-day' go in as a feature?

It doesn't for wind explicitly at least, solar it does with the solar elevation and azimuth.

**Comment by jacobbieker** [comment]
> Could the model be trained with a weighting on the first few hours, to help reduce the error in the earl forecast horizons?

Yeah, I think its pretty easy to do that

**Comment by peterdudfield** [comment]
I had a quick look at the normalized nwp variables and 'prate' stood out abit

```
| ---- | --- | ---- |
| hcc | 1.4295 |-0.9376
| lcc |-1.1843 |-1.1843
|mcc | 1.7643 | tensor(-0.8646)
| prate
| tensor(9.3231)
| tensor(-0.3168)
| sde
| tensor(-0.0887)
| tensor(-0.0887)
| sr
| tensor(4.4933)
| tensor(0.6274)
| t2m
| tensor(4.6202)
| tensor(-1.3150)
| tcc
| tensor(0.7871)
| tensor(-1.8804)
| u10
| tensor(0.3539)
| tensor(-1.2574| 
u100
| tensor(0.4887)
| tensor(-1.5100)
| u200
| tensor(0.3803)
| tensor(-1.6375)
| v10
| tensor(0.3623)
| tensor(-1.1152)
| v100
| tensor(0.6133)
| tensor(-1.4429)
| v200
| tensor(0.4583)
| tensor(-1.5155)
```

**Comment by peterdudfield** [comment]
Ive found somehting funny in the normalize part of ocf_datapipes. Mean and std seem to be different lenght that then xr_data its wokring with  
https://github.com/openclimatefix/ocf_datapipes/blob/main/ocf_datapipes/transform/xarray/normalize.py#L55

**Comment by jacobbieker** [comment]
Yep, I can look at some of the training examples. The prate shouldn't be that high, I don't think. If so, probably should be normalizing a few of those variables by different numbers.

**Comment by peterdudfield** [comment]
- [ ] Check variables names are ok
 - [ ] update lon/ lat position
 - [ ] Check normalize variables are the same
 - [ ] Check live and backtes ecmwf are in 0.1 degrees
 - [ ]

**Comment by jacobbieker** [comment]
Yeah, they are the same ones right now being used,  I didn't recalculate for india, forgot to do that

**Comment by peterdudfield** [comment]
> Yeah, they are the same ones right now being used, I didn't recalculate for india, forgot to do that

Thats ok, as long as they are the same for the moment. Might be something for the future though

**Comment by jacobbieker** [comment]
Backtest lat/lon are in 0.05 degrees increments

**Comment by jacobbieker** [comment]
Yeah, that probably sounds good, the 0.05 is already an interpolation, as the native model resolution is only 0.1 degrees

**Comment by peterdudfield** [comment]
ah, how come 0.05 is used in training? not the 0.1 native model?

**Comment by jacobbieker** [comment]
That's how we saved the data to disk for some reason. @devsjc might have a better idea as to why? But I'm pretty sure the forecast model's native resolution is 0.1 (around ~9km at the equator)

**Comment by jacobbieker** [comment]
It is how it appears in our backtest data

**Comment by peterdudfield** [comment]
could you confirm in each batch what the max and min lat and longs are?

**Comment by peterdudfield** [comment]
This looks like this makes a big difference 
from
![Screenshot 2024-03-04 at 17 53 21](https://github.com/openclimatefix/india-forecast-app/assets/34686298/98ac8459-ade2-4b1c-9e98-ef1c7912bb30)

to
![Screenshot 2024-03-04 at 17 53 48](https://github.com/openclimatefix/india-forecast-app/assets/34686298/14f5a471-7ca6-4c4a-b8c4-131988b206a5)

**Comment by jacobbieker** [comment]
> This looks like this makes a big difference from ![Screenshot 2024-03-04 at 17 53 21](https://private-user-images.githubusercontent.com/34686298/309853214-98ac8459-ade2-4b1c-9e98-ef1c7912bb30.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk1NzUxNjMsIm5iZiI6MTcwOTU3NDg2MywicGF0aCI6Ii8zNDY4NjI5OC8zMDk4NTMyMTQtOThhYzg0NTktYWRlMi00YjFjLTllOTgtZWYxYzc5MTJiYjMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA0VDE3NTQyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAxMWNkNTc2MDZhOWIxYjQ2ZWMzMDg3MzUzMTdhYTdhMzY2ZWZjNTc4MjYxZjFhMzYwYTA2N2Y4OWNkMWFjODQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.efAL-BawQJK3PHky_1ReLf0yFIG1XMt0deW5a0ipbJM)
> 
> to ![Screenshot 2024-03-04 at 17 53 48](https://private-user-images.githubusercontent.com/34686298/309853194-14f5a471-7ca6-4c4a-b8c4-131988b206a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk1NzUxNjMsIm5iZiI6MTcwOTU3NDg2MywicGF0aCI6Ii8zNDY4NjI5OC8zMDk4NTMxOTQtMTRmNWE0NzEtN2NhNi00YzRhLWI4YzQtMTMxOTg4YjIwNmE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA0VDE3NTQyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiZDZhMGQ1MTA4ODc2NGYxMWEwYzgyZjhiYWU4NjM5YzQyYzExMWFkYWE4M2U5ZDk0MGNmNjEzZGM0MDFhZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5NpWeIrPyXJjCSMQaz-FbzV9255eNfVu0q5laQyWmFU)

Is that the interpolation?

Here is all the lat/lons:
```
Latitude: [27.25 27.2  27.15 27.1  27.05 27.   26.95 26.9  26.85 26.8  26.75 26.7
 26.65 26.6  26.55 26.5  26.45 26.4  26.35 26.3  26.25 26.2  26.15 26.1
 26.05 26.   25.95 25.9  25.85 25.8  25.75 25.7 ]
Longitude: [73.85 73.9  73.95 74.   74.05 74.1  74.15 74.2  74.25 74.3  74.35 74.4
 74.45 74.5  74.55 74.6  74.65 74.7  74.75 74.8  74.85 74.9  74.95 75.
 75.05 75.1  75.15 75.2  75.25 75.3  75.35 75.4 ]
```

**Comment by peterdudfield** [comment]
thank you

**Comment by peterdudfield** [comment]
Notices a big shift in change in the forecast yesterday
<img width="768" alt="Screenshot 2024-03-06 at 07 36 02" src="https://github.com/openclimatefix/india-forecast-app/assets/34686298/46aef66c-6234-4293-963b-cf3beb432dc3">
<img width="726" alt="Screenshot 2024-03-06 at 07 35 44" src="https://github.com/openclimatefix/india-forecast-app/assets/34686298/bf56b6a2-d4bd-4886-9e25-0288079e9502">

This was becasue we got some new NWP files. Something to investigate I think

**Comment by peterdudfield** [comment]
@jacobbieker how much time lag do you add to the ECMWF data when training? It looks like in real like we get init runs from 00:00 and 12:00 and they appear about 6 or 7 hours later

**Comment by jacobbieker** [comment]
Training I think the delay is 3 hours, same as for UKV. So another thing to probably change then to 7 hours?

**Comment by jacobbieker** [comment]
This https://github.com/openclimatefix/ocf_datapipes/pull/284 will fix the normalization, and rescale the training examples to be 0.1 degrees, if its smaller than that. The config for training has also been updated to use a 7 hour delay. Once that PR is merged, can make new batches with most of the things in this thread fixed then I think.

**Comment by peterdudfield** [comment]
`sr` also does this

**Comment by dfulu** [comment]
For future reference I checked this for the UK training and production data and it was the same

**Comment by peterdudfield** [comment]
Also seen a very similar behaviour when deploying PVnet

**Comment by peterdudfield** [comment]
What I struggle is at the moment, the NWPs say its windy
![Screenshot 2024-03-13 at 08 48 04](https://github.com/openclimatefix/india-forecast-app/assets/34686298/add39017-289c-40d7-b65e-dd22e4f33bbd)
But the forecasts are low (real generation is high)
![Screenshot 2024-03-13 at 08 48 10](https://github.com/openclimatefix/india-forecast-app/assets/34686298/a13798ce-0da1-405b-ad02-bc9d39d64c7e)

**Comment by jacobbieker** [comment]
Hmmm, that might be partly because of the normalization, with the very large numbers for prate, sr, etc. compared to the wind values, the wind values might be getting ignored in the model. I can try training one that only has the wind variables, as they seem to be normalized better in the current dataset? It should force a better correlation between them.

**Comment by jacobbieker** [comment]
I was thinking just the u and v of the wind speed, as the other ones are not as tied to generation, and the u and v ones seem to be normalized better than the other ones.

**Comment by peterdudfield** [comment]
@jacobbieker do you mind confirming the max and min lat long, now we are using 0.1 for PV

In production currently they are
- lat max: 28
- lat min: 24.9
- lon max: 76.1
- lon min: 73

**Comment by jacobbieker** [comment]
The PV ones are nearly the same, have the same lat min and max, lon max is 76.2, lon min 73.1

**Comment by peterdudfield** [comment]
Ive changed the long in the database from 75.639 to 75.65 which seems to now get the correct longitude

**Comment by peterdudfield** [comment]
![Screenshot 2024-03-20 at 09 51 32](https://github.com/openclimatefix/india-forecast-app/assets/34686298/ad38c443-7f8c-41b0-b819-4ff72b00f41c)
![Screenshot 2024-03-20 at 09 51 09](https://github.com/openclimatefix/india-forecast-app/assets/34686298/35a12c39-8cbc-452e-ae92-7a4ca56d88ea)

@jacobbieker just looking at the NWP data. Looks like lat=27, lon=70 which is where some of the wind farms is super important data. Could this be included in the batch and ML model?

**Comment by jacobbieker** [comment]
Yes, it is in the latest batches now. New models will be trained soon with it, once https://github.com/openclimatefix/PVNet/pull/165 gets merged

**Comment by peterdudfield** [comment]
Another piece of evide that 27, 70 is a useful lcoation that should be included

![Screenshot 2024-03-22 at 12 37 02](https://github.com/openclimatefix/india-forecast-app/assets/34686298/117d0f32-1ce6-43c2-a7f1-f4e82f3afd07)
![Screenshot 2024-03-22 at 12 36 43](https://github.com/openclimatefix/india-forecast-app/assets/34686298/54b0e496-9c23-491e-a771-605ef4b87b3b)

**Comment by peterdudfield** [comment]
Ive set the centers to be (wind on top)

<img width="367" alt="Screenshot 2024-03-25 at 13 21 17" src="https://github.com/openclimatefix/india-forecast-app/assets/34686298/a9ebb6b9-a13c-47a1-9dcf-40eaf28c38a4">

**Comment by peterdudfield** [comment]
Thanks, sorry wind was on the bottom. Interesting, ok I will try locally and check that the correct grid is used. Thats the most important thing right?

**Comment by jacobbieker** [comment]
Yeah, the grid is the important one, we don't directly pass the lat longs to the model, so it only matters that the grids match

**Comment by dfulu** [comment]
I've been looking into the forecast performance. Here's a graph of the results including the model version number at the bottom. This is the rolling 15 minute horizon and across the month it got an MAE 240 MW (i.e. 7.11% capacity).  I've set the limits on the y-axis on the top plot so the highest value is the capacity (currently 3,381,000 KW)

![download (55)](https://github.com/openclimatefix/india-forecast-app/assets/41546094/6bf2fb95-099e-4b1f-b69a-aa4ed080e874)

**Comment by dfulu** [comment]
This is the same graph but for 24 hour horizon

To me it looks like things have been okay since v0.1.37

![download (56)](https://github.com/openclimatefix/india-forecast-app/assets/41546094/647f4829-a3de-48b2-aa7d-8461f4f7710e)
